src/main/java/edu/cornell/kfs/fp/batch/service/impl/AccountingXmlDocumentDownloadAttachmentServiceImpl.java

```diff
@@ -0,0 +1,128 @@
+import java.util.Collection;
+
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.client.Invocation.Builder;
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.io.IOUtils;
+import org.kuali.kfs.krad.bo.Attachment;
+import org.kuali.kfs.krad.document.Document;
+import org.kuali.kfs.krad.service.AttachmentService;
+import org.springframework.beans.factory.DisposableBean;
+
+import edu.cornell.kfs.fp.batch.service.AccountingXmlDocumentDownloadAttachmentService;
+import edu.cornell.kfs.fp.batch.xml.AccountingXmlDocumentBackupLink;
+import edu.cornell.kfs.sys.businessobject.WebServiceCredential;
+import edu.cornell.kfs.sys.service.WebServiceCredentialService;
+import edu.cornell.kfs.sys.service.impl.DisposableClientServiceImplBase;
+
+public class AccountingXmlDocumentDownloadAttachmentServiceImpl extends DisposableClientServiceImplBase implements AccountingXmlDocumentDownloadAttachmentService, DisposableBean {
+    private static final org.apache.log4j.Logger LOG = org.apache.log4j.Logger.getLogger(AccountingXmlDocumentDownloadAttachmentServiceImpl.class);
+
+    protected AttachmentService attachmentService;
+    protected WebServiceCredentialService webServiceCredentialService;
+
+    @Override
+    public Attachment createAttachmentFromBackupLink(Document document,
+            AccountingXmlDocumentBackupLink accountingXmlDocumentBackupLink) throws IOException {
+        File formFile = downloadFile(accountingXmlDocumentBackupLink);
+        
+        if (formFile != null) {
+            String uploadFileName = accountingXmlDocumentBackupLink.getFileName();
+            String mimeType = URLConnection.guessContentTypeFromName(formFile.getName());
+            int fileSize = (int) formFile.length();
+            String attachmentType = null;
+            
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("createAttachmentFromBackupLink, uploadFileName: " + uploadFileName + " mimeType: " + mimeType
+                    + " fileSize: " + fileSize);
+            }
+            InputStream fileContents = new FileInputStream(formFile);
+            Attachment attachment = attachmentService.createAttachment(document, uploadFileName, mimeType, fileSize,
+                    fileContents, attachmentType);
+            formFile.delete();
+            return attachment;
+        } else {
+            LOG.error("createAttachmentFromBackupLink, the form file is NULL");
+        }
+        
+        throw new IOException ("Unable to download attachment: " + accountingXmlDocumentBackupLink.getLinkUrl());
+    }
+
+    protected File downloadFile(AccountingXmlDocumentBackupLink accountingXmlDocumentBackupLink) {
+        Collection<WebServiceCredential> creds = getWebServiceCredtials(accountingXmlDocumentBackupLink);
+        File formFile = null;
+        try { 
+            Invocation request = buildClientRequest(accountingXmlDocumentBackupLink.getLinkUrl(), creds);
```

Comment:
Should we add some kind of check to make sure the URL is pointing to a legit location, and that the credential group corresponds to what is expected for that URL? Otherwise, we could end up sending web credentials to malicious endpoints, or could send the wrong set of credentials to the endpoint.

Reply:
interesting idea.  Perhaps we should standardize the group code needs to be part of the domain.  For example instead of AWS-Bill, we could use the group code "kfsaws-support.cd.cucloud.net" and look for the group code in the backup link URL.  Do you think that would meet this need?

Reply:
If not, what are your thoughts on this?

Reply:
Having the group code named as the domain would be one way of doing it. Another possibility would be to create some kind of mapping from domains to group codes. The latter could be maintained using a parameter or, alternatively, through the use of other web credential objects with the domain name as part or all of the key.

Reply:
Daniela and I talked about this, and we are going to add another credential key "credential_base_URL#" and have the value be the valid domain.